mtp/mtp_fs.cpp

```diff
@@ -4,336 +4,320 @@
-        // TODO: invalid handle
+        log_error("[%u]: stat error %s: %d", static_cast<unsigned int>(handle), filename, errno);
         return -1;
     }
 
-    LOG("[%u]: Get info for: %s", (unsigned int)handle, filename);
-
-    if (stat(abspath(fs->ROOT, filename), &statbuf)==0) {
-        memset(info, 0, sizeof(mtp_object_info_t));
-        info->storage_id = 0x00010001;
-        info->created = 1580371617;
-        info->modified = 1580371617;
-        info->format_code = ext_to_format_code(filename);
-        info->size = (uint64_t)statbuf.st_size;
-        *(uint32_t*)(info->uuid) = handle;
-        strncpy(info->filename, filename, 64);
-        return 0;
-    } else {
-        LOG("[%u]: Stat error: %s", (unsigned int)handle, filename);
-    }
+    int fs_rename(void *arg, uint32_t handle, const char *new_name)
+    {
+        const struct mtp_fs *fs = static_cast<struct mtp_fs *>(arg);
+        const char *filename    = mtp_db_get(fs->db, handle);
+        int status;
 
-    return -1;
-}
+        if (filename == nullptr) {
+            log_error("[%u]: filename is nullptr", static_cast<unsigned>(handle));
+            return -1;
+        }
 
-static int fs_rename(void *arg, uint32_t handle, const char *new_name)
-{
-    struct mtp_fs *fs = (struct mtp_fs*)arg;
-    int status;
-    const char *filename = mtp_db_get(fs->db, handle);
+        status = rename(filename, new_name);
+        if (status == 0) {
+            mtp_db_update(fs->db, handle, new_name);
+            log_debug("[%u]: rename: %s -> %s", static_cast<unsigned>(handle), filename, new_name);
+        }
+        else {
+            log_error("[%u]: rename: %s -> %s FAILED", static_cast<unsigned>(handle), filename, new_name);
+        }
+        return status;
+    }
 
-    if (!filename)
-        return -1;
+    int fs_create(void *arg, const mtp_object_info_t *info, uint32_t *handle)
+    {
+        struct mtp_fs *fs = static_cast<struct mtp_fs *>(arg);
+        uint32_t new_handle;
+        const auto absolutePath = std::string(fs->root) + std::string("/") + std::string(info->filename);
 
-    status = rename(filename, new_name);
-    if (status==0) {
-        mtp_db_update(fs->db, handle, new_name);
-        LOG("[%u]: Rename: %s -> %s", (unsigned int)handle, filename, new_name);
-    } else {
-        LOG("[%u]: Rename: %s -> %s FAILED", (unsigned int) handle, filename, new_name);
+        if (const auto freeSpace = get_free_space(arg); freeSpace < info->size) {
+            log_error("There is not enough space for file %s (%llu < %llu)", info->filename, freeSpace, info->size);
+            return -1;
+        }
+        new_handle = mtp_db_add(fs->db, info->filename);
+        if (new_handle == 0) {
+            log_error("Map is full. Can't create: %s", info->filename);
+            return -1;
+        }
+
+        fs->file = std::fopen(absolutePath.c_str(), "w");
+        if (fs->file == nullptr) {
+            log_error("Failed to open %s", info->filename);
```

Comment:
What with the product of the above _mtp_db_add()_, then?

Reply:
Added removing the file from the database if `fopen` has failed

Reply:
Now it's correct. By the way - as this is C++, you could use some RAII to simplify this resource management. If sticking to C, there's a good "goto cleanup" idiom: https://stackoverflow.com/questions/788903/valid-use-of-goto-for-error-management-in-c